src/AspNet.Security.OAuth.Shopify/ShopifyAuthenticationHandler.cs

```diff
@@ -0,0 +1,218 @@
+                throw new Exception(msg);
+            }
+
+            var shopName = properties.Items[ShopifyAuthenticationDefaults.ShopNameAuthenticationProperty];
+            var uri = string.Format(Options.AuthorizationEndpoint, shopName);
+
+            // Get the permission scope, which can either be set in options or overridden in AuthenticationProperties.
+            var scope = properties.Items.ContainsKey(ShopifyAuthenticationDefaults.ShopScopeAuthenticationProperty) ? 
+                properties.Items[ShopifyAuthenticationDefaults.ShopScopeAuthenticationProperty] : 
+                FormatScope();
+
+            var url = QueryHelpers.AddQueryString(uri, new Dictionary<string, string>
+            {
+                ["client_id"] = Options.ClientId,
+                ["scope"] = scope,
+                ["redirect_uri"] = redirectUri,
+                ["state"] = Options.StateDataFormat.Protect(properties)
+            });
+
+            // If we're requesting a per-user, online only, token, add the grant_options query param.
+            if (properties.Items.ContainsKey(ShopifyAuthenticationDefaults.GrantOptionsAuthenticationProperty))
+            {
+                var grantOptions = properties.Items[ShopifyAuthenticationDefaults.GrantOptionsAuthenticationProperty];
+                if (grantOptions != null &&
+                    grantOptions.Equals(ShopifyAuthenticationDefaults.PerUserAuthenticationPropertyValue))
+                {
+                    url = QueryHelpers.AddQueryString(url, "grant_options[]",
+                        ShopifyAuthenticationDefaults.PerUserAuthenticationPropertyValue);
+                }
+            }
+
+            return url;
+        }
+
+        /// <inheritdoc />
+        protected override async Task<OAuthTokenResponse> ExchangeCodeAsync(
+            [NotNull] string code, 
+            [NotNull] string redirectUri)
+        {
+            string shopDns;
+
+            try
+            {
+                var shop = Context.Request.Query["shop"];
+                var state = Context.Request.Query["state"];
+
+                // Shop name must end with myshopify.com
+                if (!shop.ToString().EndsWith(".myshopify.com"))
+                {
+                    throw new Exception("Unexpected query string.");    
+                }
+
+                // Strip out the "myshopify.com" suffix
+                shopDns = shop.ToString().Split('.').First();
+
+                // Verify that the shop name encoded in "state" matches the shop name we used to
+                // request the token. This probably isn't necessary, but it's an easy extra verification.
+                var z = Options.StateDataFormat.Unprotect(state);
+                if (!z.Items[ShopifyAuthenticationDefaults.ShopNameAuthenticationProperty]
+                    .Equals(shopDns, StringComparison.InvariantCultureIgnoreCase))
+                {
+                    throw new Exception("Unexpected query string");
```

Comment:
Tell the integrating developer _why_ it's "unexpected"? Maybe explicitly state that the shop is mismatched?